From 87e3d0ccd299ca01e7381ca403da7e814be2868b Mon Sep 17 00:00:00 2001
From: Seth Jennings <sjenning@redhat.com>
Date: Wed, 22 Oct 2014 16:05:44 -0400
Subject: [edac] sb_edac: avoid INTERNAL ERROR message in EDAC with unspecified channel

Message-id: <1413993944-28121-1-git-send-email-sjenning@redhat.com>
Patchwork-id: 98067
O-Subject: [RHEL7.2 PATCH] sb_edac: avoid INTERNAL ERROR message in EDAC with unspecified channel
Bugzilla: 1056990
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Aristeu Rozanski <aris@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1056990
Upstream: 351fc4a99d49fde63fe5ab7412beb35c40d27269
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=8149612

  Intel IA32 SDM Table 15-14 defines channel 0xf as 'not specified', but
  EDAC doesn't know about this and returns and INTERNAL ERROR when the
  channel is greater than NUM_CHANNELS:

  kernel: [ 1538.886456] CPU 0: Machine Check Exception: 0 Bank 1: 940000000000009f
  kernel: [ 1538.886669] TSC 2bc68b22e7e812 ADDR 46dae7000 MISC 0 PROCESSOR 0:306e4 TIME 1390414572 SOCKET 0 APIC 0
  kernel: [ 1538.971948] EDAC MC1: INTERNAL ERROR: channel value is out of range (15 >= 4)
  kernel: [ 1538.972203] EDAC MC1: 0 CE memory read error on unknown memory (slot:0 page:0x46dae7 offset:0x0 grain:0 syndrome:0x0 -  area:DRAM err_code:0000:009f socket:1 channel_mask:1 rank:0)

  This commit changes sb_edac to forward a channel of -1 to EDAC if the
  channel is not specified.  edac_mc_handle_error() sets the channel to -1
  internally after the error message anyway, so this commit should have no
  effect other than avoiding the INTERNAL ERROR message when the channel
  is not specified.

  Signed-off-by: Seth Jennings <sjenning@redhat.com>
  Signed-off-by: Mauro Carvalho Chehab <mchehab@osg.samsung.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/edac/sb_edac.c b/drivers/edac/sb_edac.c
index e06dc40..2477400 100644
--- a/drivers/edac/sb_edac.c
+++ b/drivers/edac/sb_edac.c
@@ -283,8 +283,9 @@ static const u32 correrrthrsld[] = {
  * sbridge structs
  */
 
-#define NUM_CHANNELS 4
-#define MAX_DIMMS 3  /* Max DIMMS per channel */
+#define NUM_CHANNELS  4
+#define MAX_DIMMS  3 /* Max DIMMS per channel */
+#define CHANNEL_UNSPECIFIED 0xf /* Intel IA32 SDM 15-14 */
 
 enum type {
  SANDY_BRIDGE,
@@ -1991,6 +1992,9 @@ static void sbridge_mce_output_error(struct mem_ctl_info *mci,
 
  /* FIXME: need support for channel mask */
 
+ if (channel == CHANNEL_UNSPECIFIED)
+  channel = -1;
+
  /* Call the helper to output message */
  edac_mc_handle_error(tp_event, mci, core_err_cnt,
         m->addr >> PAGE_SHIFT, m->addr & ~PAGE_MASK, 0,
-- 
1.7.1